databroker/databroker.go

```diff
@@ -1,174 +1,199 @@
-// Package databroker contains the databroker service.
+// Package databroker is a Pomerium service that handles the storage of data in Pomerium.
+// It communicates over gRPC with other Pomerium services and can be configured to use a
+// number of different backend databroker stores.
 package databroker
 
 import (
 	"context"
 	"fmt"
+	"net"
+	"time"
 
+	"github.com/rs/zerolog"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
 	oteltrace "go.opentelemetry.io/otel/trace"
-	"google.golang.org/protobuf/types/known/emptypb"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
 
 	"github.com/pomerium/pomerium/config"
 	"github.com/pomerium/pomerium/internal/atomicutil"
-	"github.com/pomerium/pomerium/internal/databroker"
+	"github.com/pomerium/pomerium/internal/events"
 	"github.com/pomerium/pomerium/internal/log"
-	databrokerpb "github.com/pomerium/pomerium/pkg/grpc/databroker"
-	registrypb "github.com/pomerium/pomerium/pkg/grpc/registry"
+	"github.com/pomerium/pomerium/internal/version"
+	"github.com/pomerium/pomerium/pkg/cryptutil"
+	"github.com/pomerium/pomerium/pkg/envoy/files"
+	"github.com/pomerium/pomerium/pkg/grpc/databroker"
+	"github.com/pomerium/pomerium/pkg/grpc/registry"
 	"github.com/pomerium/pomerium/pkg/grpcutil"
+	"github.com/pomerium/pomerium/pkg/identity"
+	"github.com/pomerium/pomerium/pkg/identity/manager"
+	"github.com/pomerium/pomerium/pkg/telemetry/trace"
 )
 
-// A dataBrokerServer implements the data broker service interface.
-type dataBrokerServer struct {
-	server    *databroker.Server
-	sharedKey *atomicutil.Value[[]byte]
-}
-
-// newDataBrokerServer creates a new databroker service server.
-func newDataBrokerServer(ctx context.Context, tracerProvider oteltrace.TracerProvider, cfg *config.Config) (*dataBrokerServer, error) {
-	srv := &dataBrokerServer{
-		sharedKey: atomicutil.NewValue([]byte{}),
-	}
+// DataBroker represents the databroker service.
+type DataBroker struct {
+	cfg         *databrokerConfig
+	srv         *server
+	identityMgr *manager.Manager
+	eventsMgr   *events.Manager
 
-	opts, err := srv.getOptions(cfg)
-	if err != nil {
-		return nil, err
-	}
-
-	srv.server = databroker.New(ctx, tracerProvider, opts...)
-	srv.setKey(cfg)
-	return srv, nil
+	localListener       net.Listener
+	localGRPCServer     *grpc.Server
+	localGRPCConnection *grpc.ClientConn
+	sharedKey           *atomicutil.Value[[]byte]
+	tracerProvider      oteltrace.TracerProvider
+	tracer              oteltrace.Tracer
 }
 
-// OnConfigChange updates the underlying databroker server whenever configuration is changed.
-func (srv *dataBrokerServer) OnConfigChange(ctx context.Context, cfg *config.Config) {
-	opts, err := srv.getOptions(cfg)
+// New creates a new databroker service.
+func New(ctx context.Context, cfg *config.Config, eventsMgr *events.Manager, options ...Option) (*DataBroker, error) {
+	localListener, err := net.Listen("tcp", "127.0.0.1:0")
```

Comment:
Is this new or refactored from somewhere else?

Reply:
This should just be refactored. I renamed the variable `c` to `d`. The `databroker` used to be called `cache` and there were still some remnants of that in the code.